packages/workers-shared/asset-server-worker/src/index.ts

```diff
@@ -1,38 +1,62 @@
+import { WorkerEntrypoint } from "cloudflare:workers";
 import { AssetsManifest } from "./assets-manifest";
+import {
+	InternalServerErrorResponse,
+	MethodNotAllowedResponse,
+	NotFoundResponse,
+	OkResponse,
+} from "./responses";
+import { getAdditionalHeaders, getMergedHeaders } from "./utils/headers";
+import { getAssetWithMetadataFromKV } from "./utils/kv";
 
-interface Env {
-	/**
-	 * ASSETS_MANIFEST is a pipeline binding to an ArrayBuffer containing the
-	 * binary-encoded site manifest
-	 */
-	ASSETS_MANIFEST: ArrayBuffer;
-	/**
-	 * ASSETS_KV_NAMESPACE is a pipeline binding to the KV namespace that the
-	 *  assets are in.
-	 */
-	ASSETS_KV_NAMESPACE: KVNamespace;
-}
-
-export default {
-	async fetch(request: Request, env: Env) {
-		const { ASSETS_MANIFEST, ASSETS_KV_NAMESPACE } = env;
+export default class extends WorkerEntrypoint<Env> {
+	async fetch(request: Request) {
+		if (request.method.toLowerCase() !== "get") {
+			return new MethodNotAllowedResponse();
+		}
 
-		const url = new URL(request.url);
-		const { pathname } = url;
+		try {
+			return this.handleRequest(request);
+		} catch (err) {
+			return new InternalServerErrorResponse(err);
+		}
+	}
 
-		const assetsManifest = new AssetsManifest(ASSETS_MANIFEST);
-		const assetKey = await assetsManifest.get(pathname);
-		if (!assetKey) {
-			return new Response("Not Found", { status: 404 });
+	async handleRequest(request: Request) {
+		const assetEntry = await this.getAssetEntry(request);
+		if (!assetEntry) {
+			return new NotFoundResponse();
 		}
 
-		const content = await ASSETS_KV_NAMESPACE.get(assetKey);
-		if (!content) {
+		const assetResponse = await getAssetWithMetadataFromKV(
+			this.env.ASSETS_KV_NAMESPACE,
+			assetEntry
+		);
+
+		if (!assetResponse || !assetResponse.value) {
 			throw new Error(
-				`Requested asset ${assetKey} exists in the asset manifest but not in the KV namespace.`
+				`Requested asset ${assetEntry} exists in the asset manifest but not in the KV namespace.`
 			);
 		}
 
-		return new Response(content);
-	},
-};
+		const { value: assetContent, metadata: assetMetadata } = assetResponse;
+		const additionalHeaders = getAdditionalHeaders(
+			assetEntry,
+			assetMetadata,
+			request
+		);
+		const headers = getMergedHeaders(request.headers, additionalHeaders);
+
+		return new OkResponse(assetContent, { headers });
+	}
+
+	private async getAssetEntry(request: Request) {
+		const url = new URL(request.url);
+		let { pathname } = url;
+
+		const assetsManifest = new AssetsManifest(this.env.ASSETS_MANIFEST);
+		pathname = globalThis.decodeURIComponent(pathname);
```

Comment:
not sure if this should be decoded because the assetManifest uses encoded pathnames?
Might be worth adding a test for url encoding with a funky pathname. Not high priority though.

Reply:
ok, so funky path names def don't work, with or without this decoding. Let's have a look together, and I'll add the proper fix/tests in a follow-up PR. Tracking in WC-2597